crates/executor/src/executor.rs

```diff
@@ -17,7 +18,157 @@ use revm::{
 use std::collections::BTreeMap;
 
 /// Main block executor
-pub struct Executor {}
+pub struct Executor<'a, DB>
+where
+    DB: StateProvider,
+{
+    chain_spec: &'a ChainSpec,
+    evm: EVM<&'a mut SubState<DB>>,
+}
+
+impl<'a, DB> Executor<'a, DB>
+where
+    DB: StateProvider,
+{
+    fn new(chain_spec: &'a ChainSpec, db: &'a mut SubState<DB>) -> Self {
+        let mut evm = EVM::new();
+        evm.database(db);
+        Executor { chain_spec, evm }
+    }
+
+    fn recover_signers(
+        &self,
+        body: &[TransactionSigned],
+        signers: Option<Vec<Address>>,
+    ) -> Result<Vec<Address>, Error> {
+        if let Some(signers) = signers {
+            if body.len() == signers.len() {
+                Ok(signers)
+            } else {
+                Err(Error::SignerRecoveryError)
+            }
+        } else {
+            body.iter().map(|tx| tx.recover_signer().ok_or(Error::SignerRecoveryError)).collect()
+        }
+    }
```

Comment:
I don't think this should be part of the interface. We can do this lazily, i.e. if looking up the recovered sender fails in the `execute` function we emit a warning and recover in-place.

```suggestion

```